File: 2024/09/9.rb
#!/usr/bin/env ruby

def checksum(disk)
  sum = 0
  disk.each_with_index do |chunk, index|
    sum += chunk * index if chunk != -1
  end
  sum
end

def compact(disk, whole_chunk = false)
  new_disk = disk.dup
  chunk_index = new_disk.size
  while chunk_index > 0
    chunk_index = chunk_index.pred
    next if new_disk[chunk_index] == -1
    chunk_size = 1
    chunk = new_disk[chunk_index]

    if whole_chunk
      # Find size and beginning of chunk
      while new_disk[chunk_index - 1] == chunk do
        chunk_index = chunk_index.pred
        chunk_size = chunk_size.succ
      end
      break if chunk_index == 0
    end

    space_index = 0
    while space_index < chunk_index
      bad_index = (space_index+chunk_size-1).downto(space_index).first { |i| new_disk[i] != -1 }

      if bad_index.nil?
      #if new_disk[space_index, chunk_size].all?(-1) # (space_index...(space_index+chunk_size)).all? { |i| new_disk[i] == -1 }
        (0...chunk_size).each do |offset|
          new_disk[space_index + offset] = chunk
          new_disk[chunk_index + offset] = -1
        end
        break
      end
      space_index = bad_index.succ
    end

  end
  new_disk
end

disk = []
$stdin.readline.chomp.chars.each_with_index do |size, index|
  if index % 2 == 0
    disk += [index/2] * size.to_i
  else
    disk += [-1] * size.to_i
  end
end

new_disk = compact(disk)
puts checksum(new_disk)

new_disk = compact(disk, true)
puts checksum(new_disk)

File: 2024/09/9.go
package main

import (
	"bufio"
	"os"
	"slices"
)

func checksum(disk []int) int {
	sum := 0
	for i, fileId := range disk {
		if fileId != -1 {
			sum += (fileId * i)
		}
	}
	return sum
}

func emptySpace(disk []int, location, size int) bool {
	for i := 0; i < size; i++ {
		if disk[location+i] != -1 {
			return false
		}
	}
	return true
}

func compact(disk []int, whole_chunk bool) []int {
	newDisk := make([]int, len(disk))
	copy(newDisk, disk)
	for chunkIndex := len(newDisk) - 1; chunkIndex > 0; chunkIndex-- {
		if newDisk[chunkIndex] == -1 {
			continue
		}
		chunkSize := 1
		fileId := newDisk[chunkIndex]

		if whole_chunk {
			for chunkIndex > 0 && newDisk[chunkIndex-1] == fileId {
				chunkSize++
				chunkIndex--
			}
			if chunkIndex == 0 {
				break
			}
		}

		for spaceIndex := range chunkIndex - 1 {
			if emptySpace(newDisk, spaceIndex, chunkSize) {
				for offset := range chunkSize {
					newDisk[spaceIndex+offset] = fileId
					newDisk[chunkIndex+offset] = -1
				}
				break
			}
		}
	}
	return newDisk
}

func parseInput() []int {
	disk := []int{}
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Scan()
	for i, d := range scanner.Text() {
		size := int(d - '0')
		chunk := make([]int, size)
		var fileId int

		if i%2 == 0 {
			fileId = i / 2
		} else {
			fileId = -1
		}

		for l := range size {
			chunk[l] = fileId
		}
		disk = slices.Concat(disk, chunk)
	}
	return disk
}

func main() {
	disk := parseInput()
	newDisk := compact(disk, false)
	println(checksum(newDisk))

	newDisk = compact(disk, true)
	println(checksum(newDisk))
}
